### PR TITLE
improve distinction between initializers and non-initializers

### PR DESCRIPTION
**Pull request by rmudambi**
### Description

- *Category*: refactor
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5380

### Changes and notes
- Replace `producer` argument with optional `initializer` argument
-  Add `is_initializer` property to ResourceGroup
- Add `is_initialized` property to Resource which is True for Columns and
  NullResources and False otherwise.

### Testing
Tests pass
Added a few new tests

### CODE REVIEW

File: src/vivarium/framework/resource/group.py [hunk 1]
```diff
@@ -22,42 +25,70 @@ class ResourceGroup:
...
     def type(self) -> str:
         """The type of resource produced by this resource group's producer."""
-        return self._resources[0].resource_type
+        return self._type
 
     @property
     def names(self) -> list[str]:
         """The long names (including type) of all resources in this group."""
-        return [resource.resource_id for resource in self._resources]
+        return list(self._resources.keys())
```
Comment (stevebachmeier) [inline src/vivarium/framework/resource/group.py:72]: nit: don't need the `.keys()`

File: src/vivarium/framework/resource/group.py [hunk 2]
```diff
@@ -22,42 +25,70 @@ class ResourceGroup:
...
 
     @property
-    def producer(self) -> Any:
-        """The method or object that produces this group of resources."""
-        return self._producer
+    def initializer(self) -> Callable[[SimulantData], None]:
+        """The method that initializes this group of resources."""
+        # todo see if we can call is_initializer instead
+        if self._initializer is None:
+            raise ResourceError("This resource group does not have an initializer.")
```
Comment (stevebachmeier) [inline src/vivarium/framework/resource/group.py:79]: Do you foresee this ever being a problem? Like if iterating through all initializers or something like that?

File: src/vivarium/framework/resource/group.py [hunk 3]
```diff
@@ -22,42 +25,70 @@ class ResourceGroup:
...
     def names(self) -> list[str]:
         """The long names (including type) of all resources in this group."""
-        return [resource.resource_id for resource in self._resources]
+        return list(self._resources.keys())
 
     @property
-    def producer(self) -> Any:
-        """The method or object that produces this group of resources."""
-        return self._producer
+    def initializer(self) -> Callable[[SimulantData], None]:
```
Comment (stevebachmeier) [inline src/vivarium/framework/resource/group.py:75]: Is the name change just to get away from overloading the term?

File: src/vivarium/framework/resource/manager.py [hunk 4]
```diff
@@ -73,47 +74,52 @@ def sorted_nodes(self) -> list[ResourceGroup]:
     def setup(self, builder: Builder) -> None:
         self.logger = builder.logging.get_logger(self.name)
 
-    # TODO [MIC-5380]: Refactor add_resources for better type hinting
```
Comment (stevebachmeier) [inline src/vivarium/framework/resource/manager.py:76]: 👍

File: src/vivarium/framework/resource/manager.py [hunk 5]
```diff
@@ -73,47 +74,52 @@ def sorted_nodes(self) -> list[ResourceGroup]:
...
-
-        for resource in resource_group:
-            if resource in self._resource_group_map:
-                other_producer = self._resource_group_map[resource].producer
+        resource_group = self._get_resource_group(resources, dependencies, initializer)
+
+        for resource_id in resource_group:
+            if resource_id in self._resource_group_map:
+                other_resource = self._resource_group_map[resource_id]
+                if resource_group.is_initializer:
```
Comment (stevebachmeier) [inline src/vivarium/framework/resource/manager.py:107]: This added logic seems like it is a bugfix? Were we previously able to have a non-initializing resource that would result in an unwanted raise?

File: src/vivarium/framework/resource/resource.py [hunk 6]
```diff
@@ -17,16 +17,33 @@ def resource_id(self) -> str:
         """The long name of the resource, including the type."""
         return f"{self.resource_type}.{self.name}"
 
+    # TODO [MIC_TBD]: make this an abstract method when support for old
```
Comment (stevebachmeier) [inline src/vivarium/framework/resource/resource.py:20]: nit: don't forget to create this ticket and add MIC #

File: src/vivarium/framework/resource/resource.py [hunk 7]
```diff
@@ -17,16 +17,33 @@ def resource_id(self) -> str:
...
+    #  requirements specification is dropped
+    @property
+    def is_initialized(self) -> bool:
+        """Return True if the resource needs to be initialized."""
+        return False
```
Comment (stevebachmeier) [inline src/vivarium/framework/resource/resource.py:25]: I guess I don't understand the `is_initialized` logic. How can this always be false for any Resource (besides Null an Columns)?

File: src/vivarium/framework/resource/group.py (hunk 3 above)
Comment (rmudambi) [reply src/vivarium/framework/resource/group.py:75]: I thought it was more consistent terminology. We already talk about simulant initializers elsewhere, and that's what these are.

File: src/vivarium/framework/resource/resource.py (hunk 7 above)
Comment (rmudambi) [reply src/vivarium/framework/resource/resource.py:25]: Initializers are always a component's `on_initialize_simulants` method. These are iterated over during simulation initialization (and are the entire reason this resource management system is required). Pipelines and RandomnessStreams don't need to be initialized - they just require certain resources to be present in order for them to be called.

File: src/vivarium/framework/resource/group.py (hunk 2 above)
Comment (rmudambi) [reply src/vivarium/framework/resource/group.py:79]: This is just so we can type hint this as not being optional

File: src/vivarium/framework/resource/manager.py (hunk 5 above)
Comment (rmudambi) [reply src/vivarium/framework/resource/manager.py:107]: This is to do with the fact that now not all resources will have an initializer (whereas preciously they all have producers). I just had an idea about how to improve this by passing the component that I will try out and maybe put up in a new PR.

**Review by stevebachmeier — APPROVED** [review]

**Review by albrja — APPROVED** [review]

**Pull request merged**